### DISCUSSION TITLE
When moving tabs around in the tree, always move all children with it

### DISCUSSION THREAD

**Discussion by airliftbluegill** [Able-Bear-Software/tab-nodes-tree#71] [Ideas / Feature Request]
Currently, if I move a tab with children, the children get "promoted" and only the one single tab moves.

This is unexpected to me, compared to how I'd expect a tree-like UI to behave. If I'm moving a parent node, I would expect all children to follow it and stay preserved within that parent.

It can be worked around by selecting every single node under the parent, but there's another unexpected behavior: If a child is collapsed, then the children within it won't move with the rest. So I first must make sure all children are expanded, and then select all, and then move.

Due to this, I usually don't like moving around parent nodes unless I really need to. I might miss a collapsed node or not get everything selected, and then mess up my tree.

I haven't yet ran into a use case where I want to move only the parent and "promote" the children left behind. If I needed to do that, my first inclination would probably be to move the child out first, then move the parent to where I want.

**Comment by zeroarst** [comment]
@airliftbluegill Thanks for your feedback! [A ticket](https://github.com/Able-Bear-Software/tab-nodes-tree/issues/57) is already in place to allow moving an entire collapsed subtree, which will address the issue where collapsed child nodes don’t move with the rest. This fix is available in version 1.4.1.

In my view, moving a collapsed subtree along with all its children makes sense. However, I don’t agree with moving the entire subtree without either selecting all or collapsing it first. If I implemented that, what if someone only wanted to move the parent node?

**Reply by airliftbluegill** [reply]
That would be great to at least fix the collapsed subtree issue. Then I could collapse a node first, then move the whole thing easier.

In terms of "what if someone only wanted to move the parent node", I'll start by saying that I won't get too hung up on it, and I can understand if you'd rather not change it. 😄 Being able to collapse and move would at least be a great improvement.

But here is my line of thinking:

I had trouble thinking of a scenario where I'd want to do that (only move the parent). So far, it's always been what I didn't want to happen and now I have to fix the children. I honestly thought it was a bug until I realized what was going on.

I think of nodes similar to folders or something. In a file explorer, there might be a very rare case where I'd want to move children up a level, but then I'd just select all the children and move them up. It's harder, but that's okay because it's a very rare use case. I definitely wouldn't expect that moving the folder would cause the folder to be emptied and all the items within it went up a level. I know tabs aren't folders though, so it's not a perfect analogy. 

Another example are bookmarks in a browser. If I move a bookmarks folder, all the contents will move with it. If I want all the children to go up a level, I'm going to need to move all the children first (which is ok because it's very rare). Another example would be Confluence pages, if you move a page, all child pages move with it. Though again I know these are not exactly the same concept, and these are all for storing long-term data rather than open tabs. Just trying to think through examples I've seen of parent/child concepts in other software and if anything else has the "only the parent moves" concept by default. And how much of a surprise this could be to other users.

**Reply by zeroarst** [reply]
I think your argument is valid, and I do see others requesting this feature. 

I’m considering adding a new option on the Options page that allows users to opt in to moving a parent node along with all its children. Additionally, pressing the Shift or Ctrl key while moving could move only the parent node. This behavior would be reversed if the option is not enabled: by default, only the parent node would move, but holding Shift or Ctrl would move all its children. I believe this strikes a balanced approach.

**Reply by stefan123t** [reply]
I would consider making the folder / bookmarks metaphor the new default and allow to disable / change it to move individual nodes using the Shift / Ctrl qualifier instead. As @airliftbluegill the common folder metaphor is to move the complete subtrees / children with the top node, With Ctrl you can usually select multiple single nodes and with Shift a range of additional nodes.

I am not so sure about the qualifiers used when moving a selection from one folder to the other. Under Linux Mint the nemo qualifiers are: 
* Move + Shift (apparently no special action), 
* Move + Ctrl `+` (**Copy** the nodes even in the same folder with extension (copy)), 
* Move + Ctrl + Shift (create a **Link** to the node in the destination)
* Move + Alt `?` (**Ask** for the action to be executed) options are
   * Move here (default with no qualifier)
   * Copy here (Ctrl as qualifier)
   * Link here (Ctrl + Shift as qualifier)

So IMHO your best bet or assignment would be Move + Shift to move only the parent node and the first child to become the new patron. :slightly_smiling_face:

**Reply by stefan123t** [reply]
@zeroarst thanks for implementing #57 for collapsed subtrees.
As I have argued above the normal behaviour / metaphor is to move the whole subtree of folders / tabs.
The special case using e.g. Shift + Move would be to detach a single node / set of nodes from the tree and move it somewhere else.
Here the behaviour for nodes under the moved away node(s) has to be taken special care of.

Actually I have been moving some tabs around recently and ended up with a situation where the tabs were not in Tab Nodes Tree anymore. 
Though they were still in the Chrome Tab Bar on top and I could somehow re-select them there. 
I could however not reach them from Tab Nodes Tree anymore.
Only after moving them to a new / separate Window (and later back to the original Window) from the top Chrome Tab Bar I could again manage them using Tab Nodes Tree.
IMHO this is another reason for not managing only the selected node(s) but rather manage the whole subtree by default.
Thanks for listening to your users and working on this rather cool and exceedingly useful plugin!

**Reply by zeroarst** [reply]
Hi @stefan123t
Providing an option for users to move a subtree or a parent node is in my plan. I was thinking about adding this feature in version 1.5.0 but ended up being too busy at work. 😔 I don't want to delay the release further. I will create a ticket for this soon.

I am currently transitioning this account to another account @Tab-Nodes-Tree, but GitHub has suspended that account. I'm not sure why. 😟

Regarding the second issue you raised, there might be a bug. I apologize for the inconvenience. If you could provide details to help me reproduce the problem, I can fix it as soon as possible. For now, if you are still experiencing the issue, you can click "Reload tree" in the context menu (right-click), which will sync to native tabs and rebuild the tree. This should resolve the missing tab nodes problem.

**Reply by stefan123t** [reply]
@zeroarst thanks for considering. I will try to reproduce the second issue.

I have noticed a third "issue" with subtrees:

![image](https://github.com/user-attachments/assets/c1fd4934-5f72-4b12-a223-76bd7f089b1d)

i.e. when closing the top most node all child nodes are moved up one level.

![image](https://github.com/user-attachments/assets/74e51126-c6f3-4156-b830-0cafb695d966)

This somehow defeats the purpose of the folders / trees. 
Especially when having multiple elements on the next / first level below the deleted node.

In Firefox / tree-style-tabs 

![image](https://github.com/user-attachments/assets/f86c7cda-3b90-4043-b61e-f863792d45ef)

the parent is closed / removed and the first remaining child inherits the position / pops up to the top of the subtree.

![image](https://github.com/user-attachments/assets/4f320f49-108d-4229-92ed-27de6a34b5c2)

That way the overall tree structure is maintained.

**Reply by zeroarst** [reply]
As I remember there is option in Firefox TST when parent node is removed, whether the first child becomes the parent node. This will be another feature. I will create a ticket for it. I am still trying to recover my account.

**Comment by Tab-Nodes-Tree** [comment]
#87

**Comment by Tab-Nodes-Tree** [comment]
Hey @stefan123t I am working on #87.

In below case, after close TN-1, TN-2 will be become the parent tab node.
Before closing TN-1:
```
TN-1
╰ TN-2
╰ TN-3
╰ TN-4
╰ TN-5
```
After closed TN-1, TN-2 is promoted to a new parent tab node.
```
TN-2
╰ TN-3
╰ TN-4
╰ TN-5
```

But for below case, what is your expectation?
Before closing TN-1:
```
TN-1
╰ TN-2
    ╰ TN-3
╰ TN-4
╰ TN-5
```

What happen after close TN-1?
TN-2 become parent tab node and TN-4 and TN-5 will be its children same as TN-3?
```
TN-2
╰ TN-3
╰ TN-4
╰ TN-5
```

or  all children tab nodes move 1 level up without promoting first child tab node as a new parent tab node.
```
TN-2
╰ TN-3
TN-4
TN-5
```

**Reply by stefan123t** [reply]
@zeroarst that is indeed a dilemma. I would prefer that TN-2 and TN-3 move up a step, but the overall structure stays within the folder. But I can see that this point of view is very much debatable and may need to be configured if you want to leave the decision to the users.

**Comment by Biep** [comment]
Just for info:
With folders, Move+Shift typically forces a move, even between file systems.  Qua implementation, this means a copy followed by a delete of the original.
Translated to TNT, this might mean that when moving tabs between windows, normal moves might copy, whereas Move+Shift would cause a true move (which might be implemented as a creation in the new window followed by a deletion of the original).
